lib/pack_index.ml

```diff
@@ -224,3 +224,312 @@ let add buf ?level:_ t =
+      end
+      else if fo_idx > 0 then begin
+        Mstruct.shift buf ((fo_idx - 1) * 4);
+        let sz0 = Int32.to_int (Mstruct.get_be_uint32 buf) in
+        let sz1 = Int32.to_int (Mstruct.get_be_uint32 buf) in
+        sz0, sz1 - sz0
+      end 
+      else
+        failwith "Pack_index.c#get_sha1_idx"
+    in
+    try
+      let ofs = sha1s_ofs + (sz0 * 20) in
+      self#scan_sha1s ~scan_thresh fo_idx sz0 ofs n sha1
+    with
+      Idx_found i -> 
+      Log.debug "c#get_sha1_idx: found:%d" i;
+      Some i
+
+  method private create_ofs64_tbl =
+    Log.debug "c#create_ofs64_tbl";
+    let buf = Mstruct.of_bigarray ~off:offsets_ofs ~len:n_sha1s_4 ba in
+    let count = ref 0 in
+    for i = 0 to n_sha1s - 1 do
+      begin
+        match Mstruct.get_uint8 buf land 128 with
+        | 0 -> ()
+        | _ -> 
+          let _ = Hashtbl.add ofs64_tbl i !count in
+          Log.debug "c#create_ofs64_tbl: %d -> %d" i !count;
+          incr count
+      end;
+      Mstruct.shift buf 3
+    done;
+    ofs64_size <- Some !count;
+    !count
+
+  method private get_fanout_idx sha1 =
+    let s = SHA.to_raw sha1 in
+    let fanout_idx = int_of_char s.[0] in
+    fanout_idx
+
+  (* implements binary search *)
+  method private scan_sha1s ?(scan_thresh=default_scan_thresh) fo_idx idx_ofs ofs n sha1 =
+    let short_sha = SHA.is_short sha1 in
+(*
+    Log.debug "c#scan_sha1s: fo_idx:%d idx_ofs:%d ofs:%d n:%d sha1:%s" 
+      fo_idx idx_ofs ofs n (if short_sha then "short" else "full");
+*)
+    let len = n * 20 in
+    let buf = Mstruct.of_bigarray ~off:ofs ~len ba in
+
+    if n > scan_thresh then begin
+      (*Log.debug "c#scan_sha1s: %d > scan_thresh(=%d)" n scan_thresh;*)
+      let p = n / 2 in
+      (*Log.debug "c#scan_sha1s: p:%d" p;*)
+      let len' = p * 20 in
+      Mstruct.shift buf len';
+      let s = SHA.input buf in
+
+      (*Log.debug "c#scan_sha1s: s=%s" (SHA.to_hex s);*)
+
+      if s = sha1 then begin
+        let idx = idx_ofs + p in
+        (*Log.debug "c#scan_sha1s: idx -> %d (full)" idx;*)
+        raise (Idx_found idx)
+      end
+      else if short_sha && SHA.is_prefix sha1 s then begin
+        let cur_ofs = ofs + len' in
+        let prev_ok =
+          if cur_ofs > sha1s_ofs then begin
+            Mstruct.shift buf 40;
+            not (SHA.is_prefix sha1 (SHA.input buf))
+          end
+          else
+            true
+        in
+        let next_ok = 
+          if cur_ofs < crcs_ofs - 20 then begin
+            Mstruct.shift buf 20;
+            not (SHA.is_prefix sha1 (SHA.input buf))
+          end
+          else
+            true
+        in
+        (*Log.debug "prev_ok:%B next_ok:%B" prev_ok next_ok;*)
+        if prev_ok && next_ok then begin
+          let idx = idx_ofs + p in
+          (*Log.debug "c#scan_sha1s: idx -> %d (short)" idx;*)
+          raise (Idx_found idx)
+        end
+        else
+          raise SHA.Ambiguous
+      end
+      else if SHA.lt sha1 s then
```

Comment:
@codinuum I'm reviewing that part of the code and I'm not sure to understand what you are doing here. The objects in the pack file are not ordered by SHA1 so I don't understand why this comparison would work.